django-s3-file-field-client/src/client.ts

```diff
@@ -25,11 +31,51 @@ export interface UploadResult {
   state: 'aborted' | 'successful' | 'error';
 }
 
-export default class S3FFClient {
-  protected baseUrl: string;
+class Progressable {
+  private overheadPercentage = 0.0;
+  private mainPercentageSegments: number[] = [];
+  // This value was picked arbitrarily
+  private readonly overheadProportion = 0.20;
+
+  constructor(
+    private readonly onProgress?: ProgressCallback,
+  ) {}
+
+  private updateProgress(): void {
+    const mainPercentage = this.mainPercentageSegments
+      .reduce((total, value) => total + value, 0.0);
+    const percentage = (this.overheadPercentage * this.overheadProportion) + (mainPercentage * (1 - this.overheadProportion));
 
-  constructor(baseUrl: string) {
-    this.baseUrl = baseUrl;
+    if (this.onProgress) {
+      // Don't leak the "this" context out to the callback
+      this.onProgress.call(undefined, {
+        percentage,
+      });
+    }
+  }
+
+  protected updateOverheadProgress(percentage: number) {
+    this.overheadPercentage = percentage;
+    this.updateProgress();
+  }
+
+  protected initializeMainProgress(segments: number) {
+    this.mainPercentageSegments = new Array(segments).fill(0);
+  }
+
+  protected updateMainProgress(segment: number, current: number, total: number) {
+    // This weights each segment equally, which is not as accurate, but easier to initialize
+    this.mainPercentageSegments[segment] = current / total;
+    this.updateProgress();
+  }
+}
+
+export default class S3FFClient extends Progressable{
+  constructor(
+    protected readonly baseUrl: string,
```

Comment:
Note this TypeScript feature for declaring and initializing member variables in one line. I've updated `baseUrl` to do that.